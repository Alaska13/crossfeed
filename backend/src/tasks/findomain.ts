import { connectToDatabase, Domain, Organization } from '../models';
import { spawnSync } from 'child_process';
import { readFileSync } from 'fs';
import { saveDomainToDb } from './helpers';
import { plainToClass } from 'class-transformer';
import { CommandOptions } from './ecs-client';

const OUT_PATH = 'out.txt';

export default async (commandOptions: CommandOptions) => {
  await connectToDatabase();

  const { organizationId, organizationName } = commandOptions;

  console.log("Running findomain on organization", organizationName);

  const organization = await Organization.findOne(organizationId);
  const { rootDomains } = organization!;
  
  for (let rootDomain of rootDomains) {
    let count = 0;
    const args = ['-it', rootDomain, '-u', OUT_PATH];
    console.log("Running findomain with args", args);
    spawnSync(
      'findomain',
      args,
      { stdio: 'pipe' }
    );

    const output = String(readFileSync(OUT_PATH));
    const lines = output.split('\n');
    for (const line of lines) {
      if (line == '') continue;
      const split = line.split(',');
      await saveDomainToDb(
        plainToClass(Domain, {
          name: split[0],
          ip: split[1],
          organization: organization
        })
      );
      count++;
    }
    console.log(`Findomain created/updated ${count} new domains`);
  }
};
