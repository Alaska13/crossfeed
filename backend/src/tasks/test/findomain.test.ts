import { handler as findomain } from '../findomain';
jest.mock('../helpers/getRootDomains');
jest.mock('../helpers/saveDomainsToDb');

jest.mock('fs', () => ({
  readFileSync: () =>
    `filedrop.cisa.gov,104.84.119.215\nwww.filedrop.cisa.gov,104.84.119.215`,
  unlinkSync: () => null
}));
jest.mock('child_process', () => ({
  spawnSync: () => null
}));

describe('findomain', () => {
  test('basic test', async () => {
    await findomain({
      organizationId: 'organizationId',
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });
  });
});
