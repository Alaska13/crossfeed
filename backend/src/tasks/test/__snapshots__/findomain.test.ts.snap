// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findomain basic test: helpers.saveDomainsToDb 1`] = `
Array [
  Domain {
    "ip": "104.84.119.215",
    "name": "filedrop.cisa.gov",
    "organization": Object {
      "id": "organizationId",
    },
  },
  Domain {
    "ip": "104.84.119.215",
    "name": "www.filedrop.cisa.gov",
    "organization": Object {
      "id": "organizationId",
    },
  },
]
`;
